ethcore/private-tx/src/log.rs

```diff
@@ -0,0 +1,247 @@
+	pub tx_hash: H256,
+	/// Current status of the private transaction
+	pub status: PrivateTxStatus,
+	/// Creation timestamp
+	pub creation_timestamp: u64,
+	/// List of validations
+	pub validators: Vec<ValidatorLog>,
+	/// Timestamp of the resulting public tx deployment
+	pub deployment_timestamp: Option<u64>,
+	/// Hash of the resulting public tx
+	pub public_tx_hash: Option<H256>,
+}
+
+/// Private transactions logging
+pub struct Logging {
+	logs: RwLock<HashMap<H256, TransactionLog>>,
+	logs_dir: Option<PathBuf>,
+}
+
+impl Logging {
+	/// Creates the logging object
+	pub fn new(logs_dir: Option<String>) -> Self {
+		let logging = Logging {
+			logs: RwLock::new(HashMap::new()),
+			logs_dir: logs_dir.map(|dir| PathBuf::from(dir)),
+		};
+		logging.read_logs();
+		logging
+	}
+
+	/// Retrieves log for the corresponding tx hash
+	pub fn tx_log(&self, tx_hash: H256) -> Option<TransactionLog> {
+		self.logs.read().get(&tx_hash).cloned()
+	}
+
+	/// Logs the creation of private transaction
+	pub fn private_tx_created(&self, tx_hash: H256, validators: &Vec<Address>) {
+		let mut validator_logs = Vec::new();
+		for account in validators {
+			validator_logs.push(ValidatorLog {
+				account: *account,
+				validated: false,
+				validation_timestamp: None,
+			});
+		}
+		let mut logs = self.logs.write();
+		if logs.len() > MAX_JOURNAL_LEN {
+			// Remove the oldest log
+			let mut sorted_logs = logs.values().cloned().collect::<Vec<TransactionLog>>();
+			sorted_logs.sort_by(|a, b| a.creation_timestamp.cmp(&b.creation_timestamp));
+			logs.remove(&sorted_logs[0].tx_hash);
+		}
+		logs.insert(tx_hash, TransactionLog {
+			tx_hash,
+			status: PrivateTxStatus::Created,
+			creation_timestamp: SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs(),
+			validators: validator_logs,
+			deployment_timestamp: None,
+			public_tx_hash: None,
+		});
+	}
+
+	/// Logs the obtaining of the signature for the private transaction
+	pub fn signature_added(&self, tx_hash: H256, validator: Address) {
+		let mut logs = self.logs.write();
+		if let Some(transaction_log) = logs.get_mut(&tx_hash) {
+			transaction_log.status = PrivateTxStatus::Validating;
+			if let Some(ref mut validator_log) = transaction_log.validators.iter_mut().find(|log| log.account == validator) {
+				validator_log.validated = true;
+				validator_log.validation_timestamp = Some(SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs());
+			}
+		}
+	}
+
+	/// Logs the final deployment of the resulting public transaction
+	pub fn tx_deployed(&self, tx_hash: H256, public_tx_hash: H256) {
+		let mut logs = self.logs.write();
+		if let Some(log) = logs.get_mut(&tx_hash) {
+			log.status = PrivateTxStatus::Deployed;
+			log.deployment_timestamp = Some(SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs());
+			log.public_tx_hash = Some(public_tx_hash);
+		}
+	}
+
+	fn read_logs(&self) {
```

Comment:
Should this not return a `Result`?

Reply:
Yes, result is added. As for its usage, see the discussion below